pkg/kv/kvclient/rangefeed/rangefeedcache/watcher.go

```diff
@@ -0,0 +1,334 @@
+	"github.com/cockroachdb/redact"
+)
+
+// Watcher is used to implement a consistent cache over spans of KV data
+// on top of a RangeFeed. Note that while rangefeeds offer events as they
+// happen at low latency, a consistent snapshot cannot be constructed until
+// resolved timestamp checkpoints have been received for the relevant spans.
+// The watcher internally buffers events until the complete span has been
+// resolved to some timestamp, at which point the events up to that timestamp
+// are published as an update. This will take on the order of the
+// kv.closed_timestamp.target_duration cluster setting (default 3s).
+//
+// If the buffer overflows (as dictated by the buffer limit the Watcher is
+// instantiated with), the old rangefeed is wound down and a new one
+// re-established. The client interacts with data from the RangeFeed in two
+// ways, firstly, by translating raw KVs into kvbuffer.Events, and by handling
+// a batch of such events when either the initial scan completes or the
+// frontier changes. The OnUpdateCallback which is handed a batch of updates
+// is informed whether the batch of events corresponds to a complete or
+// incremental update.
+//
+//
+// It's expected to Start-ed once. Start internally invokes Run in a retry
+// loop.
+//
+// The expectation is that the caller will use a mutex to update an underlying
+// data structure.
+//
+// NOTE (for emphasis): Update events after the initial scan published at a
+// delay corresponding to kv.closed_timestamp.target_duration (default 3s).
+// Users seeking to leverage the Updates which arrive with that delay but also
+// react to the row-level events as they arrive can hijack the translateEvent
+// function to trigger some non-blocking action.
+type Watcher struct {
+	name             redact.SafeString
+	clock            *hlc.Clock
+	rangefeedFactory *rangefeed.Factory
+	spans            []roachpb.Span
+	bufferSize       int
+	withPrevValue    bool
+
+	started int32 // accessed atomically
+
+	translateEvent TranslateEventFunc
+	onUpdate       OnUpdateFunc
+
+	lastFrontierTS hlc.Timestamp // used to assert monotonicity across rangefeed attempts
+
+	knobs TestingKnobs
+}
+
+// UpdateType is passed to an OnUpdateFunc to indicate whether
+// the set of events represents a complete update or a partial
+// update.
+type UpdateType bool
+
+const (
+	// CompleteUpdate indicates that the events represent the entirety of
+	// data in the spans.
+	CompleteUpdate UpdateType = true
+
+	// PartialUpdate indicates that the events represent an incremental update
+	// since the previous update.
+	PartialUpdate UpdateType = false
+)
+
+// TranslateEventFunc is used by the client to translate a low-level event
+// into an event for buffering. if nil is returned, the event is skipped.
+type TranslateEventFunc func(
+	context.Context, *roachpb.RangeFeedValue,
+) rangefeedbuffer.Event
+
+// OnUpdateFunc is used by the client to receive a batch of updates.
```

Comment:
[nit] Batch of updates? Rephrase to the singular form or refer to "events" in the plural.